compiler/src/dotty/tools/dotc/core/ContextOps.scala

```diff
@@ -34,14 +34,45 @@ object ContextOps:
               if (elem.name == name) return elem.sym.denot // return self
             }
             val pre = ctx.owner.thisType
-            pre.findMember(name, pre, required, excluded)
+            if (ctx.isJava) then javaFindMember(name, pre, required, excluded)
+            else pre.findMember(name, pre, required, excluded)
           }
           else // we are in the outermost context belonging to a class; self is invisible here. See inClassContext.
             ctx.owner.findMember(name, ctx.owner.thisType, required, excluded)
         else
           ctx.scope.denotsNamed(name).filterWithFlags(required, excluded).toDenot(NoPrefix)
       }
 
+    // Only invoke this when ctx.isJava == true
+    final def javaFindMember(name: Name, pre: Type, required: FlagSet = EmptyFlags, excluded: FlagSet = EmptyFlags): Denotation =
+      inContext(ctx) {
+        val preSym = pre.typeSymbol
+        val denot = pre.findMember(name, pre, required, excluded)
+        if (denot.exists || preSym.isPackageObject || !preSym.isClass) denot
+        else {
+          // In Java code, static innner classes, which we model as members of the companion object,
+          // can be referenced from an ident in a subclass or by a selection prefixed by the subclass.
+          val toSearch = if (preSym.is(Flags.Module)) then
+            if preSym.companionClass.exists then
+              preSym.companionClass.asClass.baseClasses
+            else Nil
+          else
+            preSym.asClass.baseClasses
+
+          toSearch.iterator.map { bc =>
+            val pre1 = bc.thisType
+            val found =  pre1.findMember(name,pre,required,excluded | Flags.TypeParam)
```

Comment:
There's an important difference with what Scala 2 does in https://github.com/scala/scala/pull/7671/commits/b1cb58aef66fb674386a79768629df2be3d92e7a here: Scala 2 uses `decl` which will look for declarations in the current class, whereas the code here uses `findMember` which will also look for members in base classes, so you should be able to do this call only once on either preSym or preSym.companionClass, then in the iterator.map you only need to look at the companions.

Also instead of .map(...).find(...) I think you could use .collectFirst(...)

Reply:
Sorry, I couldn’t find a proper partial function of collectFirst to replace .map(...).find(...)

Reply:
I think the solution is to replace `.map(...).find(_.exists)` with just `.find(f(_).exists)` where `f` is the current body of `map`

Reply:
Then, the result of find is one element of toSearch, not the result of f(_).
PS the computing in `iterator.map` is lazy until invoke `find`.

Reply:
you are right 😅